redwood-cli/src/main/kotlin/app/cash/redwood/generator/diffConsumingGeneration.kt

```diff
@@ -371,79 +370,28 @@ internal fun generateDiffConsumingWidget(schema: Schema, widget: Widget, host: S
+internal fun generateDiffConsumingLayoutModifiers(schema: Schema, host: Schema = schema): FileSpec {
   return FileSpec.builder(schema.widgetPackage(host), "layoutModifierSerialization")
     .apply {
       if (schema === host) {
-        addFunction(
-          FunSpec.builder("toLayoutModifier")
-            .addModifiers(INTERNAL)
-            .receiver(KotlinxSerialization.JsonElement)
-            .addParameter("json", KotlinxSerialization.Json)
-            .addParameter("mismatchHandler", WidgetProtocol.ProtocolMismatchHandler)
-            .returns(Redwood.LayoutModifier)
-            .addStatement("val array = %M", KotlinxSerialization.jsonArray)
-            .addStatement("require(array.size == 2) { \"Layout modifier JSON array length != 2: \${array.size}\" }")
-            .addStatement("")
-            .beginControlFlow(
-              "val serializer = when (val tag = array[0].%M.%M)",
-              KotlinxSerialization.jsonPrimitive,
-              KotlinxSerialization.jsonInt,
-            )
-            .apply {
-              if (schema.layoutModifiers.isEmpty()) {
-                addAnnotation(
-                  AnnotationSpec.builder(Suppress::class)
-                    .addMember("%S, %S, %S, %S", "UNUSED_PARAMETER", "UNUSED_EXPRESSION", "UNUSED_VARIABLE", "UNREACHABLE_CODE")
-                    .build(),
-                )
-              }
-              for (layoutModifier in schema.layoutModifiers) {
-                val typeName = ClassName(schema.widgetPackage(host), layoutModifier.type.simpleName!! + "Impl")
-                if (layoutModifier.properties.isEmpty()) {
-                  addStatement("%L -> return %T", layoutModifier.tag, typeName)
-                } else {
-                  addStatement("%L -> %T.serializer()", layoutModifier.tag, typeName)
-                }
-              }
-            }
-            .beginControlFlow("else ->")
-            .addStatement("mismatchHandler.onUnknownLayoutModifier(tag)")
-            .addStatement("return %T", Redwood.LayoutModifier)
-            .endControlFlow()
-            .endControlFlow()
-            .addStatement("")
-            .addStatement("val value = array[1].%M", KotlinxSerialization.jsonObject)
-            .addStatement("return json.decodeFromJsonElement(serializer, value)")
-            .build(),
-        )
+        addFunction(generateJsonArrayToLayoutModifier(schema))
+        addFunction(generateJsonElementToLayoutModifier(schema))
       }
 
       for (layoutModifier in schema.layoutModifiers) {
-        val typeName =
-          ClassName(schema.widgetPackage(host), layoutModifier.type.simpleName!! + "Impl")
+        val typeName = ClassName(schema.widgetPackage(host), layoutModifier.type.simpleName!! + "Impl")
         val typeBuilder = if (layoutModifier.properties.isEmpty()) {
           TypeSpec.objectBuilder(typeName)
         } else {
           TypeSpec.classBuilder(typeName)
-            .addModifiers(DATA)
```

Comment:
I removed the `data` class modifier here since we generate our own `equals`/`hashCode`/`toString` for layout modifiers.

Reply:
👍 . All `data` classes will have to be removed eventually since they're not suitable for public API / API evolution.